File: src/update/update_manager.py
# Standard library
from threading import Thread
from datetime import datetime
from webbrowser import open_new_tab

# App enums
from app_enums import UserPreferencesKeys

# Utils
from utils import get_github_version, has_one_month_passed

# User preferences
from user_preferences import UserPreferencesManager

# Constants
from constants import INSTALLED_VERSION, DOWNLOAD_PAGE_URL

# Components
from components.dialogs.updates import UpdateDialog

# Third party libraries
from flet import Page


class UpdateManager:
    """Manages update checks in a background thread, running once per month."""

    def __init__(self, app: Page):
        self.update_dialog = UpdateDialog(app, self._update)
        self.user_preferences_manager = UserPreferencesManager()

        self._start_background_update_check()

    def _start_background_update_check(self) -> None:
        """Starts a background thread to check for updates if needed."""

        Thread(target=self._check_for_updates_if_needed, daemon=True).start()

    def _check_for_updates_if_needed(self) -> None:
        """Checks if an update check is needed before querying the API."""

        if has_one_month_passed():
            self.check_updates()

    def _an_update_is_available(self):
        """Fetches the latest release from GitHub and updates preferences if needed."""

        latest_github_version = get_github_version()

        today_str = datetime.now().strftime("%Y-%m-%d")

        self.user_preferences_manager.set_preference(
            UserPreferencesKeys.LAST_UPDATE_CHECK, today_str
        )

        return INSTALLED_VERSION < latest_github_version

    def _notify_update(self, an_update_is_available: bool) -> None:
        """
        Displays the update dialog to notify the user about available updates.

        Args:
            an_update_is_available (bool): True if an update is available, False otherwise.
        """

        self.update_dialog.show_dialog(an_update_is_available)

    def _update(self) -> None:
        """
        Opens the download page for the latest version in a new browser tab.
        """

        open_new_tab(DOWNLOAD_PAGE_URL)

        self.update_dialog.close_dialog()

    def check_updates(self, checking_for_updates_manually=False) -> None:
        """
        Checks for available updates and notifies the user only when necessary.

        Args:
            checking_for_updates_manually (bool): True if the user manually triggered the check, False if automatic.
        """

        an_update_is_available = self._an_update_is_available()

        if an_update_is_available:
            self._notify_update(True)

        elif checking_for_updates_manually:
            self._notify_update(False)

File: lib/update/update_manager.dart
import 'package:fluikit/dialogs.dart' show FluiConfirmDialog, FluiInfoDialog;
import 'package:flutter/material.dart' show BuildContext;
import 'package:logkeeper/logkeeper.dart' show LogKeeper;
import 'package:pub_semver/pub_semver.dart' show Version;

import '/constants/version.dart' show installedVersion;
import '/enums/user_preferences.dart' show UserPreferencesKeys;
import '/l10n/app_localizations.dart' show AppLocalizations;
import '/managers/user_preferences_manager/user_preferences_manager.dart'
    show UserPreferencesManager;
import '/utils/github_utils.dart' show getGithubVersion;
import '/utils/time_utils.dart' show hasOneMonthPassed;

class UpdateManager {
  static Future<void> checkForUpdatesIfNecessary(BuildContext context) async {
    final lastCheckDateString = UserPreferencesManager.getPreference(
      UserPreferencesKeys.lastUpdateCheck,
    );

    if (hasOneMonthPassed(lastCheckDateString)) {
      await checkForUpdates(context: context, automaticCheckUpdates: true);
    }
  }

  static Future<bool> _anUpdateIsAvailable() async {
    try {
      final latestGithubVersion = await getGithubVersion();

      final installed = Version.parse(installedVersion);
      final latest = Version.parse(latestGithubVersion);

      return installed < latest;
    } catch (e) {
      LogKeeper.error('Error parsing version: $e');

      return false;
    }
  }

  static Future<void> checkForUpdates({
    required BuildContext context,
    bool automaticCheckUpdates = false,
  }) async {
    try {
      final anUpdateIsAvailable = await _anUpdateIsAvailable();

      final now = DateTime.now();
      final formattedDate =
          '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}';

      UserPreferencesManager.setPreference(
        UserPreferencesKeys.lastUpdateCheck,
        formattedDate,
      );

      if (anUpdateIsAvailable) {
        UserPreferencesManager.setPreference(
          UserPreferencesKeys.disclaimerShown,
          false,
        );
        UserPreferencesManager.setPreference(
          UserPreferencesKeys.whatsNewShown,
          false,
        );
        UserPreferencesManager.setPreference(
          UserPreferencesKeys.upgradeAvailable,
          true,
        );
      }

      if (!anUpdateIsAvailable && automaticCheckUpdates) return;

      if (!context.mounted) {
        LogKeeper.info('Context not mounted, skipping notification');

        return;
      }

      _notifyUpdate(context: context, anUpdateIsAvailable: anUpdateIsAvailable);
    } catch (e) {
      LogKeeper.error('Error checking for updates: $e');
    }
  }

  static void _update() {
    print('Updating...');
  }

  static void _notifyUpdate({
    required BuildContext context,
    required bool anUpdateIsAvailable,
  }) {
    if (anUpdateIsAvailable) {
      FluiConfirmDialog.show(
        context,
        title: AppLocalizations.of(context)!.update_available_title,
        content: AppLocalizations.of(context)!.update_available_body,
        onPressed: _update,
      );

      return;
    }

    FluiInfoDialog.show(
      context,
      title: AppLocalizations.of(context)!.updated_version_title,
      content: AppLocalizations.of(context)!.updated_version_body,
    );
  }

  static Future<bool> _isTheLatestVersion() async {
    final latestGithubVersion = await getGithubVersion();

    final installed = Version.parse(installedVersion);
    final latest = Version.parse(latestGithubVersion);

    return installed >= latest;
  }

  static Future<void> reminderUpdateIfNecessary(BuildContext context) async {
    final hasUpdated = await _isTheLatestVersion();

    if (!hasUpdated && context.mounted) {
      _notifyUpdate(context: context, anUpdateIsAvailable: true);

      return;
    }

    UserPreferencesManager.setPreference(
      UserPreferencesKeys.upgradeAvailable,
      false,
    );
  }
}
